### DISCUSSION TITLE
Support Regtest (correct way).

### DISCUSSION THREAD

**Discussion by oacdutra** [LNP-BP/LNPBPs#128] [Q&A] [answered]
HI @dr-orlovsky,

I will start contributing to a project for testing and building LN apps. This project works with bitcoin core and the majority of LN clients.

I want to extend this project to support RGB, Bifrost, and Storm protocols. But this project works only with regtest mode.

I found one impediment for this:
- [ ]  strict encode to bip32

The impediment occurs because the bip32 xpub prefix is the same on testnet/regtest, but the address prefix is not. To recover the correct network, I added magic bytes of the network in strict_encode/strict_decode methods.  

Do you have any consideration for the solution?

**Comment by dr-orlovsky** [comment, marked as answer]
Hm, I am not sure I understand what specificly prevents from using the xpub in the regtest network. Can you please provide more details?

Basically, xpub BIP32 strict encoding is exactly the same as the bitcoin BIP32 encoding, so if regtest is supported there it should be supported in RGB.

**Reply by oacdutra** [reply]
Hi @dr-orlovsky,

The `bip32::ExtendedPubKey` decode always sets network as "Testnet" when the prefix starts with "tpub", even if you set the network as signet or regtest. It is not a problem for signet because both PubKey and Address prefixes are the same in both networks. But when you use regtest, the Address prefix is not the same. 

This situation causes "wrong behaviors" in many projects, like lnp-node, descriptor-wallet, etc...

**Examples**

**1. Recover funding wallet (lnp-node)**  
Suppose we initialized the lnp-node with regtest network. When we start the `lnp-node` and run `lnp-node funds` method, the expected behavior is the _next address_ starts with the prefix **bcrt1** instead of **tb1**.

**2. List  addresses from wallet (descriptor-wallet)**  
Suppose I changed the project to support regtest in the `btc-hot derive` and `btc-cold create` methods. When we run `btc-cold address` method, the expected behavior is the list of addresses starts with the prefix **bcrt1** instead of **tb1**.

The best solution is to add the _magic number network_ in **ExtendedPubKey** and **ExtendedPrivKey** strict encode/decode methods to recover the correct network and avoid many changes in lnp-node, descriptior-wallet, and rust-bitcoin projects.

This make sense?

**Reply by dr-orlovsky** [reply]
Right, thank you for the explanation. But that is the bug of BIP-32 (its incompatibility with regtest address specs) and rust-bitcoin BIP32 implementation. Not sure we can fix it downstream in all the projects using rust-bitcoin and BIP32

**Reply by oacdutra** [reply]
I think this "simple way" to fix this in LNPBP related-projects is change the strict encoding for bip32 keys. 

Another approach is changing each project's command-line API to support regtest. However, this way requires more customizations.

What way do you prefer?

**Reply by dr-orlovsky** [reply]
You mean deviate strict encoding of xkeys from the BIP32 serialization?

It would break all created wallets with descriptor wallet and MyCitadel at least... Including mainnet ones. Seems to be too much. So probably command line is better even though it requires more work

**Reply by oacdutra** [reply]
Yes, you're right.

I will be planning how to make these changes.

Thanks for the feedback.

**Reply by dr-orlovsky** [reply]
From what I remember there is a network/chain parameter nearly in each command tool I did. So it just have to be re-used when the address are generated.
